### DISCUSSION TITLE
Terminology unification / Glossary

### DISCUSSION THREAD

**Discussion by marcosnils** [dagger/dagger#3194] [General]
Dagger is a tool that allows creating developer platforms composed of different elements like extensions, pipelines, workflows, scripts, etc. Sometimes myself and some Dagger users misuse or confuse this terms; the purpose of this discussion is lay out all the different concepts out there and map them to Dagger's counter-part. 

Let's use the `todoapp` kick-the-tires example we have here: https://github.com/dagger/todoapp/blob/cloak/scripts/kick-the-tires/index.mjs

## Pipeline:

It's a combination of different steps to produce an output. It's scope is a **single** graphql query. Taking our `todoapp` example:
```javascript
 // 2. Install yarn in a container
  const image = await client
    .request(
      gql`
        {
          core {
            image(ref: "index.docker.io/alpine") {
              exec(input: { args: ["apk", "add", "yarn", "git", "openssh"] }) {
                fs {
                  id
                }
              }
            }
          }
        }
      `
    )
    .then((result) => result.core.image.exec.fs.id);
```

This is would be a pipeline that "installs yarn in the container".

## Step: 

 A field in a `graphql` query which resolves to a build operation . Taking our `todoapp` example:
```javascript
 // 2. Install yarn in a container
  const image = await client
    .request(
      gql`
        {
          core {
            image(ref: "index.docker.io/alpine") {
              exec(input: { args: ["apk", "add", "yarn", "git", "openssh"] }) {
                fs {
                  id
                }
              }
            }
          }
        }
      `
    )
    .then((result) => result.core.image.exec.fs.id);
```

`image` and `exec` would be the steps here since those are the two gql fields that **resolve** to a build op.

## Script:

A composition of pipelines written in a scripting or programming language that generally (not required) uses the Dagger SDK to orchestrate it. In our `todoapp` example that would be the `index.mjs` file https://github.com/dagger/todoapp/blob/cloak/scripts/kick-the-tires/index.mjs

## SDK: 

All the tools needed to use dagger with a given language. That is a broad umbrella and encompasses:

- Engine - start+connect to dagger from a script or other code
- Extension Runtime - take code written in a given language and load it as an extension with an API
- Client Codegen - take a graphql api and turn it into a generated client that makes it easy to call dagger apis you depend on

In our `todoapp` example, we're importing the typescript SDK here https://github.com/dagger/todoapp/blob/cloak/scripts/kick-the-tires/index.mjs#L2 to start the engine https://github.com/dagger/todoapp/blob/cloak/scripts/kick-the-tires/index.mjs#L7 and handle client queries towards Dagger.

## Extension:

A way to extend Dagger pipelines by dynamically adding new steps which can be written in multiple programming languages. A Dagger project can import local and remote extensions, as well as defining new ones so other pipelines can use them. In our `todoapp` example we can see that this project imports some remote extensions here https://github.com/dagger/todoapp/blob/cloak/cloak.yaml#L9-L17 and also exposes itself as an extension here https://github.com/dagger/todoapp/blob/cloak/cloak.yaml#L18-L20

Are these assumptions accurate?
What are we missing?

cc @vikram-dagger @shykes @mircubed @jpadams

**Comment by sipsma** [comment]
There's an old (highly outdated at this point) attempt at defining all this here, just FYI that we should update that guide when we make new decisions here (e.g. delete it and replace it, probably somewhere else since I don't think the terminology is very discoverable from there): https://github.com/dagger/dagger/blob/cloak/docs/guides/bnzm7-writing_extensions.md#concepts

> SDK:

I would say an SDK is all the tools needed to use dagger with a given language. That is a broad umbrella and encompasses:
1. Engine - start+connect to dagger from a script or other code
2. Extension Runtime - take code written in a given language and load it as an extension with an API
3. Client Codegen - take a graphql api and turn it into a generated client that makes it easy to call dagger apis you depend on

It's debatable (and has been debated) whether we should keep using the umbrella term "SDK" or just break it down into different concepts right away. I personally lean towards keeping the umbrella term since it's much easier to just tell users "we have a Go SDK" than it is to *immediately* introduce them all the different subcategories ("we have 3 different things depending on exactly what you're trying to do"). They will have to learn the differences, but it's nicer if that's gradual as they ramp up and use dagger for different use cases.

However, no matter what the decision on that above, the best possible thing we can do is limit the difference between all these different subcategories of the SDK. I do see some opportunities to *reduce* the difference between scripts+extensions as part of the code-first schema support, but I don't yet see a way to totally eliminate them to the point where users are just using the exact same types/methods/interfaces. It's something we should keep thinking about though.

**Reply by marcosnils** [reply]
> I would say an SDK is all the tools needed to use dagger with a given language. That is a broad umbrella and encompasses:

+1. Just updated my initial comment.

**Comment by shykes** [comment]
Thanks for starting this Marcos.

> ## Pipeline:
> It's a combination of different steps to produce an output. It's scope is a **single** graphql query.

*Consensus: medium*

There are 3 proposed uses of "pipeline":

1. Strict definition (this one): one pipeline = one API call
2. Less strict definition: one pipeline = one or more API calls chained together
  - Example sentences:
    - "This pipeline installs yarn in a container"
    - "I refactored this pipeline to require only one API query instead of 3 previously"
3. Don't use it, users expect "pipeline" to mean something bigger and wider in scope, and would be surprised by this definition

At the moment my favorite option is 2 (use "pipeline" and allow one pipeline to be made of several API calls)

For reference, here is a definition from [wikipedia](https://en.wikipedia.org/wiki/Pipeline_(software)):

> In [software engineering](https://en.wikipedia.org/wiki/Software_engineering), a pipeline consists of a chain of processing elements ([processes](https://en.wikipedia.org/wiki/Process_(computing)), [threads](https://en.wikipedia.org/wiki/Thread_(computer_science)), [coroutines](https://en.wikipedia.org/wiki/Coroutine), [functions](https://en.wikipedia.org/wiki/Subroutine), etc.), arranged so that the output of each element is the input of the next; the name is by analogy to a physical [pipeline](https://en.wikipedia.org/wiki/Pipeline_transport). [...]Connecting elements into a pipeline is analogous to [function composition](https://en.wikipedia.org/wiki/Function_composition_(computer_science)).

> Narrowly speaking, a pipeline is linear and one-directional [...]. Flows with one-directional tree and [directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph) topologies behave similarly to (linear) pipelines – the lack of cycles makes them simple – and thus may be loosely referred to as "pipelines".

> ## Step:
> A field in a `graphql` query which resolves to a build operation . Taking our `todoapp` example:
> 
> ```js
>  // 2. Install yarn in a container
>   const image = await client
>     .request(
>       gql`
>         {
>           core {
>             image(ref: "index.docker.io/alpine") {
>               exec(input: { args: ["apk", "add", "yarn", "git", "openssh"] }) {
>                 fs {
>                   id
>                 }
>               }
>             }
>           }
>         }
>       `
>     )
>     .then((result) => result.core.image.exec.fs.id);
> ```
> 
> `image` and `exec` would be the steps here since those are the two gql fields that **resolve** to a build op.

*Consensus: low*

This is one of several proposed terms:

1. "step"
2. "action"
3. "node"
4. "task"

My favorites at the moment are "step" and "action".

* "step": simple, familiar. Used in a similar fashion in Jenkins and Github Actions.
  * Concern: does "step" imply sequential execution? This would be a problem since the pipeline model implies a chain of compute nodes executing in parallel
* "action": simple, familiar. Consistent with current version of Dagger. Draws parallel with Github Actions
  * Concern: are Github Actions the right comparison?
* Concern for both: what about GraphQL fields such as `File { contents }`? Is this a "step/action to read the file contents?" There is an impedance mismatch between the pipeline/step-or-action model and the graphql query model.

> ## Script:
> A composition of pipelines written in a scripting or programming language that generally (not required) uses the Dagger SDK to orchestrate it. In our `todoapp` example that would be the `index.mjs` file https://github.com/dagger/todoapp/blob/cloak/scripts/kick-the-tires/index.mjs

*Consensus: high*

This term is a bit different because it doesn't designate a feature of Dagger. It's just a common term for something our developers do.
I don't know how to use this information, but wanted to mention it. It's not like npm for example, which has a concept of a "npm script".

I think we should rephrase somehow but I'm not sure how.

> ## SDK:
> All the tools needed to use dagger with a given language. That is a broad umbrella and encompasses:
> 
> * Engine - start+connect to dagger from a script or other code
> * Extension Runtime - take code written in a given language and load it as an extension with an API
> * Client Codegen - take a graphql api and turn it into a generated client that makes it easy to call dagger apis you depend on

*Consensus: medium*

I would rephrase this to focus on the things you can do, rather than the internal architecture of an SDK:

All the tools needed to use dagger in a given platform or language. This includes platform-native APIs and tools to:

* Embed the Dagger engine in your code
* Compose and run pipelines
* Develop custom pipeline "steps/actions"

> ## Extension:
> A way to extend Dagger pipelines by dynamically adding new steps which can be written in multiple programming languages. A Dagger project can import local and remote extensions, as well as defining new ones so other pipelines can use them. In our `todoapp` example we can see that this project imports some remote extensions here https://github.com/dagger/todoapp/blob/cloak/cloak.yaml#L9-L17 and also exposes itself as an extension here https://github.com/dagger/todoapp/blob/cloak/cloak.yaml#L18-L20

*Consensus: low*

TLDR: **the meaning of "extension" changes between current cloak.yaml and proposed dagger.gql**.

* Current meaning: what you described above.
* Proposed meaning:
  * Extension: a software bundle which can be installed in a Dagger project to add new capabilities. Typically an extension includes specialized steps/actions which can be added in the project's pipelines.
  * Custom steps / actions: a custom task, developed in any language, which can be incorporated into a Dagger pipeline in any other language. Custom steps/actions can be private to a project, or installed by an extension.

The difference is subtle but important.

* In both cases "netlify", "yarn" etc are extensions providing custom steps/actions: no difference
* Big difference when developing custom steps/actions for your own project (todoapp deploy). In one meaning, you are creating an extension for yourself. In the other, you are creating custom steps/actions private to your project, without having to develop an extension
* The term "extension" has a connotation of plugin/customization/something you do later, not on day one. So the choice of terminology is important.

**Reply by sipsma** [reply]
> I would rephrase this to focus on the things you can do, rather than the internal architecture of an SDK:

I like your phrasing of the bullet points a lot better, ship it, but I also do think that we will need to decide on names for what I listed. They aren't fully internal; they are the actual names of methods/commands/etc. that users will be calling and thus directly exposed to. Having consistent naming for them across the SDKs is important.

That's a next level of detail though. Honestly at this point I'd rather go make all the updates to code-first schema and new client codegen and then decide on names for those individual parts. I suspect things will change enough there that it will be much easier to see the shape of what we build and come up with names after.

**Reply by shykes** [reply]
Yes 100% agree. I think those “semi-internal” terms are the next level of detail and the natural next step IMO is to finalize the core API for them. That will force some terminology-intertwined-with DX debates. Then from there, glossary :)

**Comment by mircubed** [comment]
@shykes @marcosnils @sipsma should we move this to a doc in the repo, so it is easier to see which things are finalized and which aren't since we can submit PRs to a specific definition?

**Reply by mircubed** [reply]
@shykes @marcosnils just checking in to see if we should move this to a doc in the repo now, so we can start to track what is finalized or not?

**Comment by mircubed** [comment]
Reminder that we need to add the terms from #3345 once we figure out where all these terms are living
